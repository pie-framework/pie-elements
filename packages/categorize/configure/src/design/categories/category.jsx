import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Card from '@material-ui/core/Card';
import InputHeader from '../input-header';
import CardActions from '@material-ui/core/CardActions';
import { DeleteButton } from '../buttons';

import PlaceHolder from './droppable-placeholder';

export class Category extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    category: PropTypes.object.isRequired,
    onChange: PropTypes.func,
    onDelete: PropTypes.func,
    onDeleteChoice: PropTypes.func,
    onAddChoice: PropTypes.func,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired
    })
  };

  static defaultProps = {};

  changeLabel = l => {
    const { category, onChange } = this.props;
    category.label = l;
    onChange(category);
  };

  render() {
    const {
      category,
      classes,
      className,
      onChange,
      onDelete,
      onDeleteChoice,
      onAddChoice,
      imageSupport
    } = this.props;

    return (
      <Card className={classNames(classes.category, className)}>
        {
          onChange && (
            <InputHeader
              label={category.label}
              onChange={this.changeLabel}
              onDelete={onDelete}
              imageSupport={imageSupport}
            />
          )
        }
        <PlaceHolder
          className={classes.placeHolder}
          choices={category.choices}
          onDeleteChoice={onDeleteChoice}
          onDropChoice={onAddChoice}
          categoryId={category.id}
        />
        {
          onDelete && (
            <CardActions className={classes.actions}>
              <DeleteButton label={'delete'} onClick={onDelete} />
            </CardActions>
          )
        }
      </Card>
    );
  }
}
const styles = theme => ({
  placeHolder: {
    minHeight: '100px'
  },
  deleteButton: {
    margin: 0
  },
  actions: {
    padding: 0,
    paddingBottom: 0,
    paddingTop: theme.spacing.unit
  },
  iconButtonRoot: {
    width: 'auto',
    height: 'auto'
  },
  header: {
    display: 'flex',
    justifyContent: 'space-between'
  },
  category: {
    padding: theme.spacing.unit,
    overflow: 'visible'
  },
  editor: {
    flex: '1',
    paddingBottom: theme.spacing.unit * 2
  }
});
export default withStyles(styles)(Category);
