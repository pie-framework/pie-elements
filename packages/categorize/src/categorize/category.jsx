import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Choice from './choice';
import PlaceHolder from './droppable-placeholder';
import { color } from '@pie-lib/render-ui';

export const CategoryType = {
  id: PropTypes.string.isRequired,
  categoryId: PropTypes.string,
};

export class Category extends React.Component {
  static propTypes = {
    ...CategoryType,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    classes: PropTypes.object.isRequired,
    onDropChoice: PropTypes.func,
    onRemoveChoice: PropTypes.func,
  };

  static defaultProps = {};

  render() {
    const {
      classes,
      className,
      choices = [],
      disabled,
      onDropChoice,
      onRemoveChoice,
      id,
      correct,
    } = this.props;

    const names = classNames(classes.category, className);
    const placeholderNames = classNames(
      classes.placeholder,
      correct === false && classes.incorrect
    );

    return (
      <div className={names}>
        <PlaceHolder
          onDropChoice={onDropChoice}
          disabled={disabled}
          className={placeholderNames}
        >
          {choices.map((c, index) => (
            <Choice
              onRemoveChoice={onRemoveChoice}
              disabled={disabled}
              key={index}
              choiceIndex={index}
              categoryId={id}
              {...c}
            />
          ))}
        </PlaceHolder>
      </div>
    );
  }
}

const styles = (theme) => ({
  incorrect: {
    border: `solid 2px ${color.incorrect()}`,
  },
  placeholder: {
    minHeight: '80px',
    padding: theme.spacing.unit / 2,
    borderRadius: theme.spacing.unit / 2,
    gridColumnGap: 0,
    gridRowGap: 0,
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
    alignItems: 'center',
    alignContent: 'flex-start'
  },
  category: {
    display: 'flex',
    flexDirection: 'column',
  },
});

export default withStyles(styles)(Category);
