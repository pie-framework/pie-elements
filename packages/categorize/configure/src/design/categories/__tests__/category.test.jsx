import { shallow } from 'enzyme';
import React from 'react';
import { Category } from '../category';

describe('category', () => {
  let w;

  let onChange = jest.fn();
  let onDelete = jest.fn();
  let onDeleteChoice = jest.fn();
  let onAddChoice = jest.fn();

  const wrapper = (extras) => {
    const defaults = {
      classes: {},
      className: 'className',
      category: {
        id: '1',
        label: 'Category title',
      },
      onChange,
      onDelete,
      onDeleteChoice,
      onAddChoice,
    };
    const props = { ...defaults, ...extras };
    return shallow(<Category {...props} />);
  };

  describe('snapshot', () => {
    it('renders with default props', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });

    it('renders without some components if no handlers are provided', () => {
      w = wrapper({
        onChange: undefined,
        onDelete: undefined,
      });
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('changeLabel', () => {
      it('calls onChange', () => {
        w = wrapper();
        w.instance().changeLabel('new label');
        expect(onChange).toBeCalledWith({ id: '1', label: 'new label' });
      });
    });
  });
});
