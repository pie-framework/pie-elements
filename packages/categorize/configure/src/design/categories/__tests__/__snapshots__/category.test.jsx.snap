// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`category snapshot renders with default props 1`] = `
<WithStyles(Card)
  className="className"
>
  <span>
    <WithStyles(InputHeader)
      disabled={false}
      label="Category title"
      onChange={[Function]}
      onDelete={[MockFunction]}
    />
  </span>
  <DropTarget(WithStyles(DroppablePlaceHolder))
    category={
      Object {
        "id": "1",
        "label": "Category title",
      }
    }
    categoryId="1"
    onDeleteChoice={[MockFunction]}
    onDropChoice={[MockFunction]}
  />
  <WithStyles(CardActions)>
    <WithStyles(Component)
      label="delete"
      onClick={[MockFunction]}
    />
  </WithStyles(CardActions)>
</WithStyles(Card)>
`;

exports[`category snapshot renders without some components if no handlers are provided 1`] = `
<WithStyles(Card)
  className="className"
>
  <span>
    <WithStyles(InputHeader)
      disabled={false}
      label="Category title"
      onChange={[Function]}
    />
  </span>
  <DropTarget(WithStyles(DroppablePlaceHolder))
    category={
      Object {
        "id": "1",
        "label": "Category title",
      }
    }
    categoryId="1"
    onDeleteChoice={[MockFunction]}
    onDropChoice={[MockFunction]}
  />
</WithStyles(Card)>
`;
